wandb/sdk/launch/runner/kubernetes_runner.py

```diff
@@ -64,11 +66,166 @@
+        try:
+            # Stream with no timeout polling for pod status updates
+            for event in self._pod_watcher.stream(
+                self.core_api.list_namespaced_pod,
+                namespace=self.namespace,
+                label_selector=self.pod_label_selector,
+            ):
+                type = event.get("type")
+                object = event.get("object")
+
+                if type == "MODIFIED":
+                    if object.status.phase == "Running":
+                        self._set_status(Status("running"))
+                if _is_preempted(object.status):
+                    self._set_status(Status("preempted"))
+                    self.stop()
+                    break
+                if _is_container_creating(object.status):
+                    self._set_status(Status("starting"))
+
+        # This can happen if the connection is lost to the Kubernetes API server
+        # and cannot be re-established.
+        except ApiException as e:
+            raise LaunchError(
+                "Exception when calling CoreV1Api->list_namespaced_pod: %s\n" % e
+            )
+
+        # This can happen if the connection is lost to the Kubernetes API server
+        # and cannot be re-established.
+        except urllib3.exceptions.ProtocolError as e:
+            if self.get_status().state in ["failed", "finished"]:
+                _logger.warning(f"Hanging monitor thread: {e}")
+                return
+            raise LaunchError(
+                "Exception when calling CoreV1Api->list_namespaced_pod: %s\n" % e
+            )
+
+    def _watch_job(self) -> None:
+        """Watch for job matching the jobname."""
+        try:
+            for event in self._job_watcher.stream(
+                self.batch_api.list_namespaced_job,
+                namespace="default",
+                field_selector=self.job_field_selector,
+            ):
+                object = event.get("object")
+                if object.status.succeeded == 1:
+                    self._set_status(Status("finished"))
+                    self.stop()
+                    break
+                elif object.status.failed is not None and object.status.failed >= 1:
+                    self._set_status(Status("failed"))
+                    self.stop()
+                    break
+
+        except ApiException as e:
+            raise LaunchError(
+                "Exception when calling CoreV1Api->list_namespaced_job: %s\n" % e
+            )
+
+        # This can happen if the connection is lost to the Kubernetes API server
+        # and cannot be re-established.
+        except urllib3.exceptions.ProtocolError as e:
+            if self.get_status().state in ["finished", "failed"]:
+                _logger.warning(f"Hanging monitor thread: {e}")
+                return
+            raise LaunchError(
+                "Exception when calling CoreV1Api->list_namespaced_pod: %s\n" % e
```

Comment:
list_namespaced_job?

Reply:
Also, I think these could be improved:
1. I don't think I've seen an exception with a `->` before.
2. Explicitly indicate which thread is failing.